service/history/historyCache.go

```diff
@@ -124,12 +130,15 @@ func (c *historyCache) getOrCreateWorkflowExecution(domainID string,
 				// TODO see issue #668, there are certain type or errors which can bypass the clear
 				context.clear()
 			}
-			context.Unlock()
+			context.locker.Unlock()
 			c.Release(key)
 		}
 	}
 
-	context.Lock()
+	if err := context.locker.Lock(ctx); err != nil {
```

Comment:
I think we also need to decrement the cache refcount here, as the caller will never call release.
@wxing1292 can you also take a look at this part of the change.